src/v2/guide/components-functional.md

```diff
@@ -0,0 +1,325 @@
+
+There are two common use-cases for functional components:
+
+* **Presentational** components: frequently-used pieces of a user interface, like our linked heading above, simple buttons, form inputs, etc.
+* **Higher-order** or **wrapper** components: components that will delegate to other components
+
+A **functional component** can use a template (since 2.5.0+) in a [single-file component](single-file-components.html):
+
+``` html
+<template functional>
+// slots, template logic, etc (v-if and v-for) are all usable here
+</template>
+
+<script>
+  export default {
+    name: 'my-component',
+    props: {
+    // ...
+    }
+  }
+</script>
+```
+
+or a [render function](render-function.html):
+
+``` js
+Vue.component('my-component', {
+  functional: true,
+  props: {
+    // ...
+  },
+  render: function (createElement, context) {
+    // ...
+  }
+})
+```
+
+> Note: in versions before 2.3.0, the `props` option is required if you wish to accept props in a functional component. In 2.3.0+ you can omit the `props` option and all attributes found on the component node will be implicitly extracted as props.
+
+ <a id="context-object"></a> Each of these approaches is explained in more detail below. In both cases, everything the component needs is passed through `context`, which is an object containing:
+
+- `props`: An object of the provided props
+- `children`: An array of the VNode children
+- `slots`: A function returning a slots object
+- `data`: The entire [data object](render-function.html#The-Data-Object-In-Depth), passed to the component as the 2nd argument of `createElement`
+- `parent`: A reference to the parent component
+- `listeners`: (2.3.0+) An object containing parent-registered event listeners. This is an alias to `data.on`
+- `injections`: (2.3.0+) if using the [`inject`](../api/#provide-inject) option, this will contain resolved injections.
+
+## Presentational Functional Components
+
+### As Single-File Components
+
+If you're using single-file components for the rest of your app, it may be convenient to use this pattern for presentational functional components too. Building them is much the same as building normal single-file components: 
+
+``` html
+<template functional>
+  <h1>
+    <a name="props.name" href="props.href">
+      <slot></slot>
+    </a>
+  </h1>
+</template>
+
+<script>
+  export default {
+    name: 'my-functional-component',
+    props: {
+      name: {
+        required: true;
+        type: String;
+      },
+      href: {
+        required: true,
+        type: String
+      }
+    }
+  }
+</script>
+```
+
+The `functional` keyword on the `<template>` element makes this a functional component. Rather than having access to the `name` and `href` props directly, we have access to the `context` object and its `props` property.
+
+This component can be used just as any standard component:
+
+```
+<my-functional-component name="anchor-1" href="#endnote-1">
+  My anchored heading
+</my-functional-component>
+```
```

Comment:
Maybe we could add some info about calling methods from the exported object (`export default {}`), in the template, like so:

> ### Accessing other properties from the functional component
> 
> ```vue
> <template>
>   <h1 :class="$options.methods.someLogicToReturnClasses(props)"><slot /></h1>
> </template>
> <script>
> export defalut {
>   methods: {
>     someLogicToReturnClasses(props) {
>       // some logic here
>      return '...'
>     },
>   }
> }
> </script>
> ```